### PR TITLE
Create keychain item with access for an additional application.

### PR DESCRIPTION
**Pull request by gabriel**
Uses SecKeychainItemCreateFromContent to add application access on create.
See "Creating a Custom Keychain Item" for more info: https://developer.apple.com/library/ios/documentation/Security/Conceptual/keychainServConcepts/03tasks/tasks.html#//apple_ref/doc/uid/TP30000897-CH205-TPXREF7

### CODE REVIEW

**Comment by gabriel** [comment]
I created a keychain.c file because it was easier to write and debug in Xcode. If you think I should make it pure go (not an intermediary c file), I could.

**Comment by akalin-keybase** [comment]
Yeah, I'd prefer it to be translated to go, just for the extra type-safety. But let me do a pass and make some comments...

File: keychain.c [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+ */
+OSStatus GHKeychainCreateItem(CFStringRef service, CFStringRef account, CFDataRef dataRef, SecAccessRef accessRef, SecKeychainItemRef *itemRef) {
+
+  SecKeychainAttribute attributes[] = {
+    { kSecServiceItemAttr, (UInt32)CFStringGetLength(service), (void *)CFStringGetCStringPtr(service, kCFStringEncodingUTF8) },
+    { kSecAccountItemAttr, (UInt32)CFStringGetLength(account), (void *)CFStringGetCStringPtr(account, kCFStringEncodingUTF8) },
+  };
+
+  SecKeychainAttributeList attributeList = {sizeof(attributes) / sizeof(attributes[0]), attributes};
+  return SecKeychainItemCreateFromContent(kSecGenericPasswordItemClass, &attributeList, (UInt32)CFDataGetLength(dataRef), CFDataGetBytePtr(dataRef), NULL, accessRef, itemRef);
```
Comment (akalin-keybase) [inline keychain.c]: isn't GenericPassword the wrong type? I'd think kSecSymmetricKeyItemClass is the one we'd use for this.

I think these items may show up differently in Keychain Access.

File: osxkeychain.go [hunk 2]
```diff
@@ -346,12 +347,19 @@ func GetAllAccountNames(serviceName string) (accountNames []string, err error) {
 
 	defer C.CFRelease(resultsRef)
 
+	// Check that the resultsRef is an array
```
Comment (akalin-keybase) [inline osxkeychain.go]: Did you run into this, or is this just being cautious? If the latter, I'd add a comment saying something like using kSecMatchLimitAll is supposed to make the type of resultsRef an array, but double-check it.

File: osxkeychain.go [hunk 3]
```diff
@@ -346,12 +347,19 @@ func GetAllAccountNames(serviceName string) (accountNames []string, err error) {
...
+	typeId := C.CFGetTypeID(resultsRef)
+	if typeId != C.CFArrayGetTypeID() {
+		err = fmt.Errorf("Invalid result type: %s", _CFStringToUTF8String(C.CFCopyTypeIDDescription(typeId)))
```
Comment (akalin-keybase) [inline osxkeychain.go]: Need to CFRelease the returned CFString

File: osxkeychain.go [hunk 4]
```diff
@@ -359,3 +367,36 @@ func GetAllAccountNames(serviceName string) (accountNames []string, err error) {
 	}
 	return
 }
+
+
+// Add a generic password keychain item with access for the specified application (and ourself).
+// This allows us to create a keychain item and allow access for an additional application or executable.
+func AddGenericPasswordWithApplication(serviceName string, accountName string, data []byte, applicationPath string) (err error) {
```
Comment (akalin-keybase) [inline osxkeychain.go]: re. the comment above, this should be something like AddKeyDataWithApplication, right?

File: osxkeychain_test.go [hunk 5]
```diff
@@ -234,3 +234,21 @@ func TestRemoveAndAddGenericPassword(t *testing.T) {
...
+	applicationPath := "/Applications/Mail.app"
+	data := []byte("testdata")
+
+	err := AddGenericPasswordWithApplication(serviceName, accountName, data, applicationPath)
+	if err != nil {
+		t.Error(err)
+	}
+
+	// Cleanup
+	err = FindAndRemoveGenericPassword(&GenericPasswordAttributes{ServiceName: serviceName, AccountName: accountName})
```
Comment (akalin-keybase) [inline osxkeychain_test.go]: Ah, I guess if you use the symmetric key type, you won't be able to use this function.

I think you can do something similar to GetAllAccountNames, though. You'd build a query dictionary with kSecClass = kSecClassSymmetricKey (or whatever), then call SecItemDelete on it.

**Comment by akalin-keybase** [comment]
Okay, did a first pass with some comments.

**Comment by gabriel** [comment]
Thanks! Yeah I was looking at symmetric key type. I'll give that a shot.

**Comment by gabriel** [comment]
So I can't get the symmetric key type to work with either SecKeychainItemCreateFromContent or SecItemAdd (with kSecAttrAccess). I was poking around https://github.com/snej/MYCrypto/blob/master/MYSymmetricKey.m to figure out the required attributes and it looks like only certain types of keys are supported.

You probably might feel strongly that GenericPassword type isn't appropriate for this use case but I think it's the safest and most straight forward approach given that keychain (symmetric and asymmetric) key items require specific inputs and capabilities that don't apply to us.

I could also move this patch into a separate library (or keybase/client) if it's something that doesn't feel appropriate here.

**Comment by akalin-keybase** [comment]
D'oh, I didn't realize the key item types required a certain format. In that case, I'm okay with using GenericPassword.

How about this:
- Change GenericPasswordAttributes.Password to be []byte, as you did before, but don't remove the code in keybase/client that encodes to base64. (You can do whatever you want with the item for the key, since that's short-lived anyway, right?)
- Add a TrustedApplicationPaths []string member to GenericPasswordAttributes.
- Change AddGenericPassword to use basically the go translation of the C code, using the strings in GenericPasswordAttributes.TrustedApplicationPaths

That'll work, right?

**Comment by gabriel** [comment]
yep sounds good

File: osxkeychain.go (hunk 2 above)
Comment (gabriel) [inline osxkeychain.go]: Yeah this is a sanity thing.. It's nice to have if we add a fetch "reference" capability in the future. I added this in when I was testing other fetch methods and thought it was useful to keep in.

**Comment by gabriel** [comment]
ok working with TrustedApplications all in go. Decided to use SecItemAdd with kSecAttrAccess (for trusted applications) instead of SecKeychainItemCreateFromContent with SecAccessRef arg, since it's a better, easier to use API, especially in cgo.

File: osxkeychain.go [hunk 6]
```diff
@@ -25,12 +25,13 @@ import (
 // All string fields must have size that fits in 32 bits. All string
 // fields except for Password must be encoded in UTF-8.
 type GenericPasswordAttributes struct {
-	ServiceName string
-	AccountName string
-	Password    string
+	ServiceName         string
+	AccountName         string
+	Password            []byte
+	TrustedApplications []string
```
Comment (akalin-keybase) [inline osxkeychain.go:36]: add "and TrustedApplications" to the comment, since it doesn't have to be utf-8

Comment (akalin-keybase) [inline osxkeychain.go:36]: Also add check32bit calls in CheckValidity for TrustedApplications

Comment (akalin-keybase) [inline osxkeychain.go:36]: Also add a comment saying that TrustedApplications refers to other programs, and that the current one is always trusted

File: osxkeychain.go [hunk 7]
```diff
@@ -108,32 +109,50 @@ func (ke keychainError) Error() string {
...
-	defer C.free(unsafe.Pointer(accountName))
+	var accountNameString C.CFStringRef
+	if accountNameString, err = _UTF8StringToCFString(attributes.AccountName); err != nil {
+		return
+	}
+	defer C.CFRelease(C.CFTypeRef(accountNameString))
 
-	password := unsafe.Pointer(C.CString(attributes.Password))
-	defer C.free(password)
+	dataBytes := C.CFTypeRef(bytesToCFData(attributes.Password))
```
Comment (akalin-keybase) [inline osxkeychain.go]: nit: i prefer to have dataBytes be of type CFData and to have CFRelease and the map value cast it (like serviceNameString)

File: osxkeychain.go (hunk 2 above)
Comment (akalin-keybase) [inline osxkeychain.go]: ah, okay. please add the comment

File: osxkeychain.go [hunk 8]
```diff
@@ -108,32 +109,50 @@ func (ke keychainError) Error() string {
...
+		defer C.CFRelease(C.CFTypeRef(access))
+		query[C.kSecAttrAccess] = C.CFTypeRef(access)
+	}
+
+	queryDict := mapToCFDictionary(query)
+	defer C.CFRelease(C.CFTypeRef(queryDict))
+
+	errCode := C.SecItemAdd(queryDict, nil)
+
+	err = newKeychainError(errCode)
```
Comment (akalin-keybase) [inline osxkeychain.go:164]: The type of the return value of FindGenericPassword should probably be consistent with GenericPasswordAttributes.Password

**Comment by akalin-keybase** [comment]
Did another pass, mostly nits. Lookin' good, though.

**Comment by gabriel** [comment]
Ok I fixed for all the feedback. Before merging I'll need to update keybase/client.

File: osxkeychain.go [hunk 9]
```diff
@@ -57,6 +63,11 @@ func (attributes *GenericPasswordAttributes) CheckValidity() error {
 	if err := check32Bit("Password", attributes.Password); err != nil {
 		return err
 	}
+	for _, trustedApplication := range attributes.TrustedApplications {
+		if err := check32BitUTF8("TrustedApplications", trustedApplication); err != nil {
```
Comment (akalin-keybase) [inline osxkeychain.go:67]: can you pass in `fmt.Sprintf("TrustedApplications[%d]", i)` for the first arg to check32BitUTF8?

**Comment by akalin-keybase** [comment]
lgtm! can you make another pull request for the keybase/client change, too? (just to make sure I didn't forget something on that side...)

**Comment by gabriel** [comment]
yeah...  https://github.com/keybase/client/pull/559

**Pull request merged**